### ISSUE TITLE
allDay event does not span multiple days

### ISSUE THREAD

**Issue by FCView** [builttoroam/device_calendar#220]
I want to be able to create an event that spans several days, and is an All Day event.

Currently, when I try to do so, it only posts to the calendar as a 1-day All Day event on the first day.

```
Event e = new Event(MyCalendarID);
e.eventId = MyEventId;
e.title = "My Title";
e.description = "My description";
e.start = new DateTime.utc(2020, 4, 1);
e.end = new DateTime.utc(2020, 5, 2);
e.allDay = true;
await _deviceCalendarPlugin.createOrUpdateEvent(e);
```

When I later call DeviceCalendarPlugin.retrieveEvents(...), I get the event back, but the end date now matches the start date (Event.start == Event.end) and it displays only on that start date on my device calendar.

This is on Android.

**Comment by brendabullorini** [comment]
My problem is the opposite.
If I put allDay = false, it creates the an all day event

**Comment by nickrandolph** [comment]
Sounds like a bug with the way that we're setting event start and end dates when it's an all day event. Feel free to open a PR with suggested fix

**Comment by voivoed** [comment]
I don't know if anyone already looked into this, but what I'm seeing is this on Android:

1) I created an all day event that goes from 08/18/20 to 08/19/20 (2 full days) on my calendar;
2) when I retrieve the event, the start date is 08/17/20 20:00:00 and end date is 08/18/20 20:00:00;
3) I see the same thing happening on events from the Holiday calendar, start/end date are short 4 hours, making them start on the previous day and end earlier on the actual date;
4) I'm on EST, which is 4 hours earlier than UTC.

Could it be that the plugin treats start/end times for all day as UTC, even though is already on my local timezone, and then converts the dates unnecessarily? That would cause what I'm seeing.

**Comment by FCView** [comment]
@voivoed , see my pull request ( #256 ) in the Example app for what is going on. Basically, when we create the event, it needs to be from midnight to midngiht. However, when we retrieve the allDay event, the Calendar Provider returns it adjusted to our local timezone, as you mentioned. So, we need to adjust it back to UTC to get the correct start and end dates.

My problem was that when using this package to create an allDay event that spans more than one day, this package was only creating it on the first day (it turns out that it was just not setting the end date). That should be fixed in my PR.

**Comment by voivoed** [comment]
@FCView that's what I had to do, I had to convert it back to get the actual dates, then just use the UTC dates as if they were local.

I created the events I needed using the actual device calendar app, my app doesn't create events yet, but may in the future, so hopefully by the time I need it to they'll have merged your PR :-)

**Comment by FCView** [comment]
I found another issue, though, that I don't think is fixable within the scope of this package. I think this is an issue with either the Android's Calendar Provider, or the way Android handles local Calendars.

When I add an allday event to a Calendar that is not tied to a synced account (I think this is Calendar.accountType == LOCAL, but I am going to look more into it tomorrow), then the allDay event can get pushed into adjacent days on the Device Calendar when the timezone is later changed.

For example, if I add a one-day allday event via this package, and then later change timezones, then the event still shows the exact same information, but it is now showing up on two days on the Calendar app (even though no information in the Event has changed and it has the exact same info as a Calendar app-created allday event). The calendar-created allDay event does not spill over into an adjacent day, but allday events created with the exact same information from this package (or more specifically, Android's CalendarContract) will spill over.

I'm using spill over, because it will spill over forward or backward depending on which direction your timezone changes.

This does not appear to be app-specific as I saw this behavior on both Google's Calendar app and Samsung's Calendar app.

For the synced calendars, I used calendars that were synced with both Google and Yahoo and did not see this behavior. Only for a Local calendar.

My users travel across timezones a lot, so this is a pretty big issue for me and I may just restrict writing to local calendars until this issue is addressed. I need to do some more digging into this, though.

**Comment by vasilich6107** [comment]
Still an issue on iOs

**Comment by vasilich6107** [comment]
Hi @voivoed 
could you check my PR #413

**Comment by voivoed** [comment]
> Hi @voivoed could you check my PR #413

Hey @vasilich6107 I don't have write access to this repository, so my review won't matter 🙂

**Comment by vasilich6107** [comment]
Thanks @voivoed for quick response

I will try to ask 
@sowens-csd or @nickrandolph 
for review

**Comment by IVLIVS-III** [comment]
@henry2man Yes, indeed. This is fixed since 4.3.0.
Thanks again for the reminder.
